app/features/user-authentication/user-authentication-actions.server.ts

```diff
@@ -53,15 +52,13 @@ export type LoginActionData = {
   errors?: FieldErrors<z.infer<typeof loginFormSchema>>;
 };
 
-async function loginHandler({
-  data,
+export const loginAction = async ({
   request,
-  t,
-}: {
-  data: z.infer<typeof loginSchema>;
-  request: Request;
-  t: TFunction;
-}) {
+}: Pick<ActionFunctionArgs, 'request'>) => {
+  const anonymousRequest = await requireAnonymous(request);
+  const data = await parseFormData(loginSchema, anonymousRequest);
+  const t = await i18next.getFixedT(anonymousRequest);
```

Comment:
These can probably be parallelized and you can use `request` for both `parseFormData` and `getFixedT`. That way, you can delete the `anonymousRequest`.